### DISCUSSION TITLE
stm32 disco kit running zephyr application from external NOR flash in XiP

### DISCUSSION THREAD

**Discussion by FRASTM** [zephyrproject-rtos/zephyr#73570] [Show and tell]
The purpose is to run a simple zephyr application on the external NOR flash, in XiP mode of a stm32 disco kit.

The application is a simple hello_word : the application can not use the external NOR itself in anycase.
The target is the stm32h750b_dk discovery kit from STMicroelectronics  (more targets to come)

The stm32H750 disco kit has a dual quad-spi flash mt25ql512ab which will be configured in **memory mapped** mode.
The memory Mapped mode allows code execution directly from the external NOR : e**X**ecute **I**n **P**lace.
The zephyr application is built and linked and flashed for the external NOR with MCUboot.
The mcuboot is always running in the internal flash memory of the stm32 mcu. The mcuboot can identify the image on the external NOR and jumps there for executing in Place.

Few steps to complete  :
have the flash_stm32_qspi supporting dual flash mode (for stm32h7 board with dual-qspi flash memories)
define the MPU to have ext_memory region access for execution

Most important is the definitions of the partitions
The internal flash (&flash0) must hold the boot partition (offset = 0x0) plus a (dummy) slot0_partition (offset = 0x10000)
The external memory must hold the **slot1** partition where the mcuboot will find binary image  to jump to.
The board DTS must select the chosen flash0 and qspi-nor-flash  controller : 
```
zephyr,flash = &flash0;
zephyr,flash-controller = &mt25ql512ab1;
```

- for MCUboot 

the zephyr, chosen are flash0 and boot_partition : 
```
chosen {

		zephyr,code-partition = &boot_partition;
};
```
kconfig  required 
```
CONFIG_BOOT_DIRECT_XIP=y
CONFIG_STM32_MEMMAP=y
```

- for the zephyr application
the zephyr, chosen is the slot1_partition : 

```
	chosen {
		zephyr,code-partition = &slot1_partition;
	};
```

- build and link

Then the mcuboot and application are built with --sysbuild (refer to https://docs.zephyrproject.org/latest/build/sysbuild/index.html ) with the following command:
`west build -p auto -b stm32h750b_dk samples/boards/stm32/hello_world_xip/ --sysbuild -- -DSB_CONFIG_BOOTLOADER_MCUBOOT=y -DCONFIG_FLASH_BASE_ADDRESS=0x90000000`

(The CONFIG_FLASH_BASE_ADDRESS gives the address of the external memory where the application will be linked.)

- download
Then flash build/mcuboot/zephyr/zephyr.bin  at 0x8000000 (internal flash address + offset of the boot_partition)
Then flash build/_your_application_/zephyr/zephyr.signed.bin at 0x90000000 (external flash address, given by CONFIG_FLASH_BASE_ADDRESS + offset of the slot1_partition)

- run
```
** Booting MCUboot v2.1.0-rc1-7-gb0cb1a475c42 ***
*** Using Zephyr OS build v3.6.0-4919-g22e2a76c3e22 ***
I: Starting Direct-XIP bootloader
I: Image 0 Primary slot: Image not found
I: Secondary slot: version=0.0.0+0
I: Image 0 loaded from the secondary slot
I: Bootloader chainload address offset: 0x0
I: Jumping to the image slot
*** Booting Zephyr OS build v3.6.0-4919-g22e2a76c3e22 ***
Hello World! from external flash  stm32h750b_dk
```

**Comment by FRASTM** [comment]
Branches are available for the stm32h750b_dk (requires the https://github.com/zephyrproject-rtos/zephyr/pull/73027)

1. for the mcuboot :https://github.com/mcu-tools/mcuboot/pull/1891
2. for the zephyr : https://github.com/FRASTM/zephyr/tree/73570

**Comment by FRASTM** [comment]
restrictions :
not availabe on stm32h747i_disco kit 
cannot power off/on after flashing until another flash operation of the external NOR (see https://github.com/zephyrproject-rtos/zephyr/discussions/56221)

**Comment by hongphucpolycrystech** [comment]
D:/Project2025/zephyrproject/SettonIO/app_mcuboot/hello-world/h750b/mcuboot/zephyr/.config.sysbuild:5: warning: attempt to assign the value 'n' to the undefined symbol BOOT_SWAP_USING_OFFSET

I don't know why compiler said undefined symbol BOOT_SWAP_USING_OFFSET. when I build with command: 
west build -p auto -d h750b -b stm32h750b_dk D:\Project2025\zephyrproject\zephyr\samples\hello_world --sysbuild -- -DSB_CONFIG_BOOTLOADER_MCUBOOT=y -DCONFIG_FLASH_BASE_ADDRESS=0x90000000
Please help
